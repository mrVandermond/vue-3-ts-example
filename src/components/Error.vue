<template>
  <div
    :class="['error', {
      'error__title_right': props.isTitleFromRight,
    }]"
  >
    <img
      src="assets/error.png"
      :class="['error-image', {
        'error-image_without-margin': props.isTitleFromRight,
      }]"
      alt="error-image"
    >

    <div
      :class="{
        'title_position_right': props.isTitleFromRight,
      }"
    >
      <p class="title-text title-text__bold">
        Сервер не отвечает
      </p>

      <p class="title-text">
        Уже работаем над этим
      </p>
    </div>
  </div>
</template>

<script lang="ts" setup>
const props = withDefaults(defineProps<{
  isTitleFromRight?: boolean;
}>(), {
  isTitleFromRight: false,
});
</script>

<style lang="scss">
.error-image {
  margin-bottom: 32px;

  &_without-margin {
    margin-bottom: 0;
  }
}
.error {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  max-width: 264px;
}
.error__title_right {
  max-width: initial;
}
.title_position_right {
  margin-left: 24px;
}
</style>
