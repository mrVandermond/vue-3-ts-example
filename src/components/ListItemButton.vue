<template>
  <div
    :class="{
      'is-open': props.isOpen,
    }"
    class="list-item-button"
  />
</template>

<script lang="ts" setup>
const props = defineProps<{
  isOpen: boolean;
}>();
</script>

<style lang="scss">
.list-item-button {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  position: relative;
  background-color: $cyan;

  &:before,
  &:after {
    content: '';
    position: absolute;
    background-color: $white;
  }

  &:before {
    width: 2px;
    height: 14px;
    top: calc(50% - 7px);
    left: calc(50% - 1px);
  }

  &:after {
    width: 14px;
    height: 2px;
    top: calc(50% - 1px);
    left: calc(50% - 7px);
  }

  &.is-open:before {
    display: none;
  }
}
</style>
