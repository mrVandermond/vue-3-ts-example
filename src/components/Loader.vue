<template>
  <img
    class="loader"
    src="assets/loader.gif"
    alt="loader-image"
  >
</template>

<script lang="ts" setup>
</script>

<style lang="scss">
.loader {
  width: 48px;
  height: 48px;
}
</style>
