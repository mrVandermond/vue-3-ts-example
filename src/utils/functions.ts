export function getItemIndexBy<T>(array: T[], item: T, field: keyof T): number {
  return array.findIndex((temp) => temp[field] === item[field]);
}
