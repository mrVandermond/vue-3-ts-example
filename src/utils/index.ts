export * as utlFuncs from './functions';
export * as utlCst from './constants';
