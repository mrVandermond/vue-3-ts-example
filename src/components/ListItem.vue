<template>
  <div
    :class="{
      internal: props.isInternal,
      'is-open': isOpen
    }"
    class="list-item"
  >
    <div
      class="list-item-header"
      @click="onClickHeader"
    >
      <ListItemButton
        :is-open="isOpen"
        class="list-item-button"
      />

      <slot name="label" />
    </div>

    <div
      :class="{
        'is-loading': props.isLoadingContent,
      }"
      class="list-item-content"
    >
      <Loader v-if="props.isLoadingContent" />

      <Error
        v-else-if="!props.isLoadingContent && props.isFailLoadingContent"
        :is-title-from-right="true"
      />

      <slot
        v-else
        name="content"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue';
import ListItemButton from '@/components/ListItemButton.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';

const isOpen = ref(false);

const props = withDefaults(defineProps<{
  isInternal?: boolean;
  isLoadingContent?: boolean;
  isFailLoadingContent?: boolean;
}>(), {
  isInternal: false,
  isLoadingContent: false,
  isFailLoadingContent: false,
});
const emit = defineEmits<{
  (event: 'click', isOpen: boolean): void;
}>();

const onClickHeader = (): void => {
  isOpen.value = !isOpen.value;
  emit('click', isOpen.value);
};
</script>

<style lang="scss">
.list-item {
  padding: 24px 0;

  &.internal {
    margin-left: 56px;
  }

  &.is-open {
    padding-bottom: 0;
  }
}
.list-item-button {
  margin-right: 24px;
}
.list-item-header {
  cursor: pointer;
  display: flex;
  align-items: center;
}
.list-item-content.is-loading {
  min-height: 320px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
}
</style>
