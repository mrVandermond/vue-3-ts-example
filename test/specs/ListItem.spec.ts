import ListItem from '@/components/ListItem.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';
import ListItemButton from '@/components/ListItemButton.vue';

import type { TWrapperFactoryReturnType } from '../utils';

import { wrapperFactory } from '../utils';

describe('ListItem component', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const getWrapper = (props: any = {}, slots: any = {}): TWrapperFactoryReturnType => wrapperFactory(ListItem, {
    props,
    slots,
  });

  it('Рендеринг на дефолтных props', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Рендеринг, когда isInternal = true', () => {
    const wrapper = getWrapper({
      isInternal: true,
    });

    expect(wrapper.element.classList).toContain('internal');
  });

  it('Рендеринг, когда isLoadingContent = true', () => {
    const wrapper = getWrapper({
      isLoadingContent: true,
    });

    expect(wrapper.find('.list-item-content').classes('is-loading')).toBe(true);
    expect(wrapper.findComponent(Loader).exists()).toBe(true);
  });

  it.each([
    true,
    false,
  ])('Рендеринг, когда isFailLoadingContent = true и isLoadingContent = %s', (expected) => {
    const wrapper = getWrapper({
      isFailLoadingContent: true,
      isLoadingContent: !expected,
    });
    const error = wrapper.findComponent(Error);

    expect(error.exists()).toBe(expected);

    if (expected) {
      expect(error.props('isTitleFromRight')).toBe(true);
    }
  });

  it('Проверка слота label', () => {
    const wrapper = getWrapper(undefined, {
      label: '<div class="slot-label"></div>',
    });

    expect(wrapper.find('.slot-label').exists()).toBe(true);
  });

  it('Проверка слота content', () => {
    const wrapper = getWrapper(undefined, {
      content: '<div class="slot-content"></div>',
    });

    expect(wrapper.find('.slot-content').exists()).toBe(true);
  });

  it('Клик по хедеру', async () => {
    const wrapper = getWrapper();
    const header = wrapper.find('.list-item-header');

    await header.trigger('click');

    expect(wrapper.classes('is-open')).toBe(true);
    expect(wrapper.findComponent(ListItemButton).props('isOpen')).toBe(true);
    expect(wrapper.emitted<boolean[]>('click')?.[0][0]).toBe(true);
  });

  it('Два клика по хедеру', async () => {
    const wrapper = getWrapper();
    const header = wrapper.find('.list-item-header');

    await header.trigger('click');
    await header.trigger('click');

    expect(wrapper.classes('is-open')).toBe(false);
    expect(wrapper.findComponent(ListItemButton).props('isOpen')).toBe(false);
    expect(wrapper.emitted<boolean[]>('click')?.[1][0]).toBe(false);
  });
});
