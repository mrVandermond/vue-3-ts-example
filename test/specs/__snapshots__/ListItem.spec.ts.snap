// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListItem component > Рендеринг на дефолтных props 1`] = `
"<div class=\\"list-item\\">
  <div class=\\"list-item-header\\">
    <list-item-button-stub isopen=\\"false\\" class=\\"list-item-button\\"></list-item-button-stub>
  </div>
  <div class=\\"list-item-content\\"></div>
</div>"
`;
